tests/Database/QueryBuilderTest.php

```diff
@@ -0,0 +1,259 @@
+        $builder->getConnection()
+            ->expects($this->once())
+            ->method('affectingStatement')
+            ->with('insert into "users" ("email", "name") values (?, ?), (?, ?) on conflict ("email") do update set "name" = "excluded"."name"', ['foo', 'bar', 'foo2', 'bar2'])
+            ->willReturn(2);
+        $result = $builder->from('users')->upsert([['email' => 'foo', 'name' => 'bar'], ['name' => 'bar2', 'email' => 'foo2']], 'email', ['name']);
+        $this->assertEquals(2, $result);
+
+        // SQLite
+        $builder = $this->getSQLiteBuilder();
+        $builder->getConnection()
+            ->expects($this->once())
+            ->method('affectingStatement')
+            ->with('insert into "users" ("email", "name") values (?, ?), (?, ?) on conflict ("email") do update set "name" = "excluded"."name"', ['foo', 'bar', 'foo2', 'bar2'])
+            ->willReturn(2);
+        $result = $builder->from('users')->upsert([['email' => 'foo', 'name' => 'bar'], ['name' => 'bar2', 'email' => 'foo2']], 'email', ['name']);
+        $this->assertEquals(2, $result);
+
+        // SQL Server
+        $builder = $this->getSqlServerBuilder();
+        $builder->getConnection()
+            ->expects($this->once())
+            ->method('affectingStatement')
+            ->with('merge [users] using (values (?, ?), (?, ?)) [laravel_source] ([email], [name]) on [laravel_source].[email] = [users].[email] when matched then update set [name] = [laravel_source].[name] when not matched then insert ([email], [name]) values ([email], [name])', ['foo', 'bar', 'foo2', 'bar2'])
+            ->willReturn(2);
+        $result = $builder->from('users')->upsert([['email' => 'foo', 'name' => 'bar'], ['name' => 'bar2', 'email' => 'foo2']], 'email', ['name']);
+        $this->assertEquals(2, $result);
+    }
+
+    protected function getConnection()
+    {
+        $connection = $this->getMockBuilder(ConnectionInterface::class)
+            ->disableOriginalConstructor()
+            ->disableOriginalClone()
+            ->disableArgumentCloning()
+            ->disallowMockingUnknownTypes()
+            ->setMethods([
+                'table',
+                'raw',
+                'selectOne',
+                'select',
+                'cursor',
+                'insert',
+                'update',
+                'delete',
+                'statement',
+                'affectingStatement',
+                'unprepared',
+                'prepareBindings',
+                'transaction',
+                'beginTransaction',
+                'commit',
+                'rollBack',
+                'transactionLevel',
+                'pretend',
+            ])
+            ->addMethods([
+                'getDatabaseName',
+            ])
+            ->getMock();
+
+        $connection->method('getDatabaseName')->willReturn('database');
+
+        return $connection;
+    }
+
+    protected function getBuilder()
```

Comment:
@LukeTowers where is this method used?

Reply:
Not sure, ask @bennothommo

Reply:
@mjauvin It's not used, but it may be in the future if we add more generic tests for the builder. I ported these methods from Laravel, to be fair.